### DISCUSSION TITLE
on_conn返回的指针地址不一样

### DISCUSSION THREAD

**Discussion by huaijinli** [Tencent/tquic#217] [General]
为什么on_conn_created和on_conn_established返回的conn 和tctx不一样，现在会crash，怀疑跟这个不一样有关系

**Comment by glime-ai** [comment]
根据您提供的信息，似乎您遇到的问题涉及到在使用TQUIC库时，`on_conn_created`和`on_conn_established`回调中的`conn`对象（或者是`tctx`，具体取决于您的上下文）不一致导致的崩溃。在TQUIC的文档和讨论中，并没有直接提及`on_conn_created`和`on_conn_established`这两个回调函数，这可能是因为这些细节特定于库的实现或者是您项目的特定封装。

不过，从一般的网络库设计和QUIC协议的实现来看，`on_conn_created`可能是在连接对象刚被创建时调用，此时连接尚未建立；而`on_conn_established`则是在连接成功建立后调用。如果`conn`或`tctx`在这两个回调中不一致，可能是因为：

1. **生命周期管理问题**：在连接建立过程中，连接对象可能被重新分配或更新，导致在不同阶段引用的对象不同。
2. **多线程或并发问题**：如果TQUIC库在处理连接时涉及到多线程或异步操作，可能存在线程安全问题或竞态条件，导致对象状态在不同回调中不一致。
3. **API使用不当**：可能是在使用TQUIC API时，某些步骤或使用方式不正确，导致了对象状态的不一致。

为了解决这个问题，您可以尝试以下步骤：

- **查阅文档**：仔细查阅TQUIC的官方文档，看是否有关于这两个回调的具体说明，以及如何正确处理连接对象。
- **检查生命周期管理**：确保`conn`或`tctx`的生命周期在整个连接过程中被正确管理，没有被意外修改或释放。
- **确保线程安全**：如果您的应用涉及到多线程，确保对`conn`或`tctx`的访问是线程安全的。
- **寻求社区帮助**：如果问题仍然存在，可以在TQUIC的GitHub讨论区或相关社区提问，可能有遇到类似问题的开发者能提供帮助。

如果您能提供更多关于如何使用TQUIC的代码示例或错误日志，可能有助于进一步诊断问题。

**Comment by iyangsj** [comment]
@huaijinli 使用的是tquic哪一个版本，能否提供复现方式或复现代码

**Reply by huaijinli** [reply]
0.8

**Reply by huaijinli** [reply]
复现也很简单就是在on_conn_established这个回调里面再调用quic_stream_new，

**Reply by iyangsj** [reply]
以https://github.com/tquic-group/tquic-example-c 项目为例，修改如下

```
$ git diff
diff --git a/simple_client.c b/simple_client.c
index e03bd8a..aca9ba4 100644
--- a/simple_client.c
+++ b/simple_client.c
@@ -47,9 +47,12 @@ struct simple_client {
 void client_on_conn_created(void *tctx, struct quic_conn_t *conn) {
     struct simple_client *client = tctx;
     client->conn = conn;
+    printf("DEBUG:     on_conn_created(): conn %p ctx %p\n", (void *)conn, (void *)tctx);
 }
 
 void client_on_conn_established(void *tctx, struct quic_conn_t *conn) {
+    printf("DEBUG: on_conn_established(): conn %p ctx %p\n", (void *)conn, (void *)tctx);
+    quic_stream_new(conn, 0, 0, false);
     const char *data = "GET /\r\n";
     quic_stream_write(conn, 0, (uint8_t *)data, strlen(data), true);
 }
```

make编译运行simple_server/simple_client:
```
./simple_server 127.0.0.1 8888
```

```
./simple_client 127.0.0.1 8443 2>&1 | grep DEBUG
```

<img width="1040" alt="企业微信截图_1aee1cf1-2e8b-4e3d-91c6-6592aef8920a" src="https://github.com/Tencent/tquic/assets/780807/23b819e9-7302-4599-a1d5-01746dbd651c">

没有复现你提到的问题。

**Reply by iyangsj** [reply]
如果方便，可以提供下相关的程序看看

**Comment by huaijinli** [comment]
quic_stream_new ，调用这个就会crash

**Comment by huaijinli** [comment]
on_conn_established 返回的tex固定是0x55，然后调用quic_stream_new jiu's就会crash
